lua/telescope/builtin/files.lua

```diff
@@ -289,134 +289,153 @@ local function prepare_match(entry, kind)
+end
 
-  local is_dir = function(value)
-    return value:sub(-1, -1) == os_sep
+local browse_folders = function(opts)
+  local data = {}
+  if vim.fn.executable "fd" == 1 then
+    local cmd = { "fd", "-t", "d" }
+    if opts.hidden then
+      table.insert(cmd, "-H")
+    end
+    if not opts.respect_gitignore then
+      table.insert(cmd, "-I")
+    end
+    return finders.new_oneshot_job(
+      cmd,
+      { entry_maker = opts.entry_maker { cwd = opts.cwd, is_fd = true }, cwd = opts.cwd }
+    )
+  else
+    scan.scan_dir(opts.cwd, {
+      hidden = opts.hidden,
+      only_dirs = true,
+      respect_gitignore = opts.respect_gitignore,
+      on_insert = function(entry)
+        table.insert(data, entry .. os_sep)
+      end,
+    })
+    table.insert(data, 1, "." .. os_sep)
+    return finders.new_table { results = data, entry_maker = opts.entry_maker { cwd = opts.cwd } }
   end
+end
 
-  opts.depth = opts.depth or 1
-  opts.cwd = opts.cwd and vim.fn.expand(opts.cwd) or vim.loop.cwd()
-  opts.new_finder = opts.new_finder
-    or function(path)
-      opts.cwd = path
-      local data = {}
-
-      scan.scan_dir(path, {
-        hidden = opts.hidden or false,
-        add_dirs = true,
-        depth = opts.depth,
-        on_insert = function(entry, typ)
-          table.insert(data, typ == "directory" and (entry .. os_sep) or entry)
-        end,
-      })
-      table.insert(data, 1, ".." .. os_sep)
-
-      local maker = function()
-        local mt = {}
-        mt.cwd = opts.cwd
-        mt.display = function(entry)
-          local hl_group
-          local display = utils.transform_path(opts, entry.value)
-          if is_dir(entry.value) then
-            display = display .. os_sep
-            if not opts.disable_devicons then
-              display = (opts.dir_icon or "") .. " " .. display
-              hl_group = "Default"
-            end
-          else
-            display, hl_group = utils.transform_devicons(entry.value, display, opts.disable_devicons)
-          end
-
-          if hl_group then
-            return display, { { { 1, 3 }, hl_group } }
-          else
-            return display
-          end
-        end
-
-        mt.__index = function(t, k)
-          local raw = rawget(mt, k)
-          if raw then
-            return raw
-          end
-
-          if k == "path" then
-            local retpath = Path:new({ t.cwd, t.value }):absolute()
-            if not vim.loop.fs_access(retpath, "R", nil) then
-              retpath = t.value
-            end
-            if is_dir(t.value) then
-              retpath = retpath .. os_sep
-            end
-            return retpath
-          end
-
-          return rawget(t, rawget({ value = 1 }, k))
-        end
-
-        return function(line)
-          local tbl = { line }
-          tbl.ordinal = Path:new(line):make_relative(opts.cwd)
-          return setmetatable(tbl, mt)
+local fb_finder = function(opts)
```

Comment:
There's an annoying issue I'll have to look into and is probably not so easily fixed.

1. Multi select
2. Search
3. Find multi selection not highlighted again

This currently happens, because this PR creates new finders eagerly in `_call`. This affects `master` in a different form.

1. Launch `file_browser`
2. Multi select
3. Go to different folder
4. Return & find multi selections gone

Ultimately, the multi selection is inadequately preserved since the entry is created anew from a new finder resulting in a new metatable for the entry, such that `picker:is_multi_selected` doesn't result in `true` for the same entry. I'll have to think about this some more to come up with an hopefully easy fix (maybe add selection as a table w/o `metatable`, if easily possible?).

Reply:
Now the entries are cached by their absolute path to transparently remain multi selections across file and folder browsers.